libvast/vast/format/csv.hpp

```diff
@@ -31,4 +43,47 @@ class writer : public format::ostream_writer {
   std::string last_layout_;
 };
 
+/// A reader for CSV data. It operates with a *selector* to determine the
+/// mapping of CSV object to the appropriate record type in the schema.
+class reader final : public single_layout_reader {
+public:
+  using super = single_layout_reader;
+
+  /// Constructs a CSV reader.
+  /// @param table_slice_type The ID for table slice type to build.
+  /// @param in The stream of CSV lines.
+  explicit reader(caf::atom_value table_slice_type,
+                  std::unique_ptr<std::istream> in = nullptr);
+
+  void reset(std::unique_ptr<std::istream> in);
+
+  caf::error schema(vast::schema sch) override;
+
+  vast::schema schema() const override;
+
+  const char* name() const override;
+
+protected:
+  caf::error read_impl(size_t max_events, size_t max_slice_size,
+                       consumer& f) override;
+
+private:
+  using iterator_type = std::string::const_iterator;
+  using parser_type = type_erased_parser<iterator_type>;
+
+  caf::optional<record_type> make_layout(const std::vector<std::string>& names);
+
+  caf::error read_header(std::string_view line);
+
+  std::unique_ptr<std::istream> input_;
+  std::unique_ptr<detail::line_range> lines_;
+  vast::schema schema_;
+  struct rec_table {
+    record_type type;
+    std::vector<std::string> sorted;
+  };
```

Comment:
We usually don't inline types like this. You can put it to the type aliases under `private:`.